third_party/move/tools/move-package-manifest/src/manifest.rs

```diff
@@ -0,0 +1,328 @@
+}
+
+/// Location of a package dependency.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum PackageLocation {
+    /// Refers to a package stored in the local file system.
+    Local { path: String },
+
+    /// Refers to a package stored in a git repository.
+    Git {
+        /// URL to the Git repository.
+        url: Url,
+        /// Optional Git revision to pin the dependency to.
+        rev: Option<String>,
+        /// Optional subdirectory within the Git repository.
+        subdir: Option<String>,
+    },
+
+    /// Refers to a package published on-chain.
+    // REVISIT: Potentially leaky abstraction -- do we still want this to be platform agnostic?
+    Aptos {
+        /// URL to the Aptos full-node connected to the network where the package is published.
+        node_url: String,
+
+        /// Address of the published package.
+        package_addr: String,
+    },
+}
+
+/***************************************************************************************************
+ * Custom Deserializer Implementations
+ *
+ **************************************************************************************************/
+impl<'de> Deserialize<'de> for Version {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let make_err = || serde::de::Error::custom("Invalid version format");
+
+        let s = String::deserialize(deserializer)?;
+        let parts: Vec<&str> = s.split('.').collect();
+        if parts.len() != 3 {
+            return Err(make_err());
+        }
+        let major = parts[0].parse().map_err(|_| make_err())?;
+        let minor = parts[1].parse().map_err(|_| make_err())?;
+        let patch = parts[2].parse().map_err(|_| make_err())?;
+
+        Ok(Version {
+            major,
+            minor,
+            patch,
+        })
+    }
+}
+
+impl<'de> Deserialize<'de> for AddressOrWildcard {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        Ok(match s.as_str() {
+            "_" => Self::Wildcard,
+            _ => Self::Numerical(
+                AccountAddress::from_str(&s)
+                    .map_err(|_| serde::de::Error::custom("Invalid account address"))?,
+            ),
+        })
+    }
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+struct RawDependency {
+    version: Option<Version>,
+
+    local: Option<String>,
+
+    git: Option<Url>,
+    rev: Option<String>,
+    subdir: Option<String>,
+
+    aptos: Option<String>,
+    address: Option<String>,
+}
+
+impl<'de> Deserialize<'de> for Dependency {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let raw = RawDependency::deserialize(deserializer)?;
+
+        let location = match (raw.local, raw.git, raw.aptos) {
```

Comment:
This match means that expressions like

`a = { local = "../a", version = "1.3.3", rev = "main" }` are valid even though `rev` isn't really doing anything here. Same for `subdir`, `address`. So we want this?

Reply:
Yeah was aware of this already but it wasn't the most straightforward to check. Finally came up with working solution though. 

It now errors on a field if associated tag name ("git", "aptos") is missing
```
error: failed to parse manifest
  |
1 | [package]
2 | name = "some_package_name"
3 | version = "0.1.2"
4 |
5 | [dependencies]
6 | a = { local = "../a", address = "0x123" }
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ redundant field "address" -- only needed for "aptos" dependencies
  |
```